{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from synbio.annotations import Location, Part\n",
    "\n",
    "import z3\n",
    "from z3helpers.definitions import *\n",
    "from z3helpers.constraints import *\n",
    "from z3helpers.utils import add_constraints\n",
    "\n",
    "# phiX174 specific stuffs\n",
    "with open(\"phiX174.pkl\", \"rb\") as handle:\n",
    "    phiX174 = pickle.load(handle)\n",
    "    \n",
    "region = Location(1155, 1521, \"FWD\")\n",
    "offset = region.start\n",
    "\n",
    "geneA = Part(seq=phiX174, name=\"A gene\", kind=\"CDS\", location=region)\n",
    "geneB = phiX174.annotations[\"B gene\"]\n",
    "\n",
    "# define z3 variables\n",
    "T = code_dict(triplet_dna_codons, AminoBitVecSort)\n",
    "dna_seq = dna_variables(region, NucleotideEnumSort)\n",
    "geneA_prot_seq = protein_variables(geneA, AminoBitVecSort)\n",
    "geneB_prot_seq = protein_variables(geneB, AminoBitVecSort)\n",
    "\n",
    "# add translation constraints\n",
    "geneA_translation_constraints = translation_constraints(\n",
    "    T, dna_seq, geneA_prot_seq, geneA.location, aminos=z3_bitvec_aminos, offset=offset)\n",
    "geneB_translation_constraints = translation_constraints(\n",
    "    T, dna_seq, geneB_prot_seq, geneB.location, \n",
    "    start_flag=True, stop_flag=True, aminos=z3_bitvec_aminos, offset=offset)\n",
    "\n",
    "# add soft constraints\n",
    "translation_obj = translates_same(geneA_prot_seq, geneA, amino_to_z3_bitvec_amino)\n",
    "translation_obj += translates_same(geneB_prot_seq, geneB, amino_to_z3_bitvec_amino)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unsat\n"
     ]
    }
   ],
   "source": [
    "# set up z3\n",
    "z3.set_param(\"verbose\", 10)\n",
    "z3.set_param(\"smt.bv.eq_axioms\", False)\n",
    "z3.set_param(\"smt.phase_caching_on\", 80000)\n",
    "solver = z3.Optimize()\n",
    "# add hard constraints\n",
    "add_constraints(solver, amino_bitvec_unary_restriction())\n",
    "add_constraints(\n",
    "    solver, RED20(T, aminos=z3_bitvec_aminos, amino_dict=amino_to_z3_bitvec_amino)\n",
    ")\n",
    "add_constraints(solver, geneA_translation_constraints)\n",
    "add_constraints(solver, geneB_translation_constraints)\n",
    "# add soft constraints\n",
    "add_constraints(solver, translation_obj, hard=False)\n",
    "# solve\n",
    "if solver.check() == z3.sat:\n",
    "    print(solver.model())\n",
    "else:\n",
    "    print(z3.unsat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some multiplexing\n",
    "Goal: probe problem complexity by testing subsets of the problem, i.e., recoding the first $N$ codons of gene B embedded in gene A."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pickle\n",
    "from typing import Optional\n",
    "import pickle\n",
    "\n",
    "from synbio.annotations import Location, Part\n",
    "from synbio.polymers import DNA\n",
    "\n",
    "import z3\n",
    "from z3helpers.definitions import *\n",
    "from z3helpers.constraints import *\n",
    "from z3helpers.utils import add_constraints\n",
    "\n",
    "# phiX174 specific stuffs\n",
    "with open(\"phiX174.pkl\", \"rb\") as handle:\n",
    "    phiX174 = pickle.load(handle)\n",
    "    \n",
    "NUM_CODONS = 30\n",
    "A_START = 1155\n",
    "A_END = 1521\n",
    "region = Location(A_START, A_END, \"FWD\")\n",
    "\n",
    "geneA = Part(seq=phiX174, name=\"A gene\", kind=\"CDS\", location=region)\n",
    "geneB = phiX174.annotations['B gene']\n",
    "\n",
    "def get_part(base_part: Part, num_codons: int, start: Optional[int] = None) -> Part:\n",
    "    base_loc = base_part.location\n",
    "    base_seq = base_part._seq_reference\n",
    "    base_name = base_part.name\n",
    "    base_kind = base_part.kind\n",
    "    \n",
    "    if start is None:\n",
    "        start = base_loc.start\n",
    "    \n",
    "    new_loc = Location(start, int(start + 3*num_codons), \"FWD\")\n",
    "    return Part(seq=base_seq, name=f\"{base_name} ({num_codons} codons)\", \n",
    "                kind=base_kind, location=new_loc)\n",
    "\n",
    "subset_A = get_part(geneA, NUM_CODONS, start=A_START)\n",
    "subset_B = get_part(geneB, NUM_CODONS)\n",
    "\n",
    "# define z3 variables\n",
    "T = code_dict(triplet_dna_codons)\n",
    "dna_seq = dna_variables(region)\n",
    "geneA_prot_seq = protein_variables(subset_A)\n",
    "geneB_prot_seq = protein_variables(subset_B)\n",
    "\n",
    "# add translation constraints\n",
    "geneA_translation_constraints = translation_constraints(\n",
    "    T, dna_seq, geneA_prot_seq, subset_A.location, offset=A_START)\n",
    "geneB_translation_constraints = translation_constraints(\n",
    "    T, dna_seq, geneB_prot_seq, subset_B.location, \n",
    "    start_flag=True, offset=A_START)\n",
    "\n",
    "# add soft constraints\n",
    "translation_obj = translates_same(geneA_prot_seq, subset_A)\n",
    "translation_obj += translates_same(geneB_prot_seq, subset_B)\n",
    "\n",
    "# set up z3\n",
    "z3.set_param(\"verbose\", 10)\n",
    "z3.set_param(\"smt.bv.eq_axioms\", False)\n",
    "z3.set_param(\"smt.phase_caching_on\", 80000)\n",
    "solver = z3.Optimize()\n",
    "# solver.set(\"timeout\", 1000)\n",
    "# add hard constraints\n",
    "add_constraints(solver, RED20(T))\n",
    "add_constraints(solver, geneA_translation_constraints)\n",
    "add_constraints(solver, geneB_translation_constraints)\n",
    "# add soft constraints\n",
    "add_constraints(solver, translation_obj, hard=False)\n",
    "# solve\n",
    "solver.check()\n",
    "m = solver.model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from z3helpers.utils import rmap\n",
    "from z3helpers.definitions import dna_to_z3nucleotide as dna_dict\n",
    "from z3helpers.analysis import objective_function\n",
    "\n",
    "from synbio.polymers import DNA\n",
    "\n",
    "def code_from_model(T, model):\n",
    "    rna = lambda dna: dna.replace(\"T\", \"U\")\n",
    "    amino = lambda codon: str(model[T[codon]])\n",
    "    \n",
    "    dict_ = {\n",
    "        rna(codon): amino(codon) for codon in T.keys()\n",
    "    }\n",
    "    \n",
    "    return Code(dict_)\n",
    "\n",
    "class AmbiguousDNA(DNA):\n",
    "    basepairing = {\n",
    "        k:v for k,v in DNA.basepairing.items()\n",
    "    }\n",
    "    basepairing['X'] = 'X'\n",
    "    \n",
    "    def alphabet(self):\n",
    "        dna_alphabet = super().alphabet()\n",
    "        dna_alphabet.append('X')\n",
    "        return dna_alphabet\n",
    "\n",
    "def dna_from_model(dna_variables, model):\n",
    "    dict_ = rmap(dna_dict)\n",
    "    \n",
    "    dna_assignments = (\n",
    "        model[var] for var in dna_seq\n",
    "    )\n",
    "    dna_str = (\n",
    "        'X' if nuc is None else dict_[nuc]\n",
    "        for nuc in dna_assignments\n",
    "    )\n",
    "    return AmbiguousDNA(\n",
    "        ''.join(dna_str)\n",
    "    )\n",
    "\n",
    "    \n",
    "def prot_from_model(prot_varibles, model):\n",
    "    amino = lambda z3aa: str(model[z3aa])\n",
    "    prot_seq = map(amino, prot_varibles)\n",
    "    \n",
    "    return Protein(\n",
    "        ''.join(prot_seq)\n",
    "    )\n",
    "\n",
    "\n",
    "# extract values\n",
    "r20 = code_from_model(T, m)\n",
    "dna = dna_from_model(dna_seq, m)\n",
    "pA = prot_from_model(geneA_prot_seq, m)\n",
    "pB = prot_from_model(geneB_prot_seq, m)\n",
    "\n",
    "# measure awesomeness\n",
    "def cost_fxn(out_seq, wt_seq, weights=None):\n",
    "    if weights is None:\n",
    "        weights = [1] * len(out_seq)\n",
    "        \n",
    "    matches = (\n",
    "        a1 == a2\n",
    "        for a1, a2 in zip(out_seq, wt_seq)\n",
    "    )\n",
    "    return sum(\n",
    "        match * weight\n",
    "        for match, weight in zip(matches, weights)\n",
    "    )\n",
    "\n",
    "cost_fxn(pA, subset_A.seq.translate()) + cost_fxn(pB, subset_B.seq.translate())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "virtual-env",
   "language": "python",
   "name": "virtual-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
